### ISSUE TITLE
`Global` doesn't respect `Send` and `Sync` on the `#[global_allocator]` type

### ISSUE THREAD

**Issue by repnop** [rust-lang/rust#85487]
I'm working on a hobby operating system and one of the things I'm doing is having allocators in userspace be defaulted to being only single-threaded, and I discovered today with some testing on the playground that the [`Global`](https://doc.rust-lang.org/alloc/alloc/struct.Global.html) allocator type has the `Send` and `Sync` autotraits applied, regardless of if the global allocator is `Send` or `Sync` (in the example below, it is neither). The expectation is the `std::thread::spawn` would error as the `Box` bound to `foo` should not impl `Send`, thus preventing the move, as the actual underlying allocator (`Foo`) isn't `Send`.

```rust
#![feature(thread_local)]

struct Foo(*mut ());

unsafe impl core::alloc::GlobalAlloc for Foo {
    unsafe fn alloc(&self, layout: core::alloc::Layout) -> *mut u8 { /* IRRELEVANT */ std::alloc::System.alloc(layout) }
    unsafe fn dealloc(&self, ptr: *mut u8, layout: core::alloc::Layout) { /* IRRELEVANT */ std::alloc::System.dealloc(ptr, layout) }
}

#[thread_local]
#[global_allocator]
static FOO: Foo = Foo(core::ptr::null_mut());

fn main() {
    let foo = Box::new(0u32);
    let _ = std::thread::spawn(move || println!("{}", foo)).join();
}
```
[playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f327b945cd2b2aed6604c83b8c2c18ab)

**Comment by Mark-Simulacrum** [comment]
I don't think the global allocator attribute is intended to be combined with the thread local attribute (it doesn't really make sense), and if that attribute isn't applied I think statics require send/sync already?

**Comment by repnop** [comment]
well the whole point is that userland applications will have a thread-local allocator for their own memory, utilizing `#[global_allocator]` so it doesn't need to specified every single time (though I guess this could be handled with an alias in my custom `std`, but seems a bit messy). If I want allocations to be `!Send`, `#[thread_local]` is the only option I have for making a global allocator, because of the requirements on normal `static`s to be `Send + Sync`, at which point every `Box` and co. will also be marked as `Send` when that's actually incorrect behavior. I know its a bit of a strange usecase, but I wasn't sure of any way to actually accomplish it :(

**Comment by Aaron1011** [comment]
Independent of the `#[global_allocator]` and `#[thread_local]` issue - the default allocator used by `Box` is always [Global](https://doc.rust-lang.org/std/alloc/struct.Global.html), *not* the type of the `#[global_allocator]` static. The `Global` type is always `Send` and `Sync`, which doesn't support your use-case.

**Comment by repnop** [comment]
But that's the thing: `Global` forwards the calls to the `#[global_allocator]` but does not have the same `Send` and `Sync` as that type, which seems wrong? If _all_ global allocators are expected to be `Send + Sync` then I would expect that to be enforced by the `GlobalAlloc` trait instead and not just by the `static` requirements.

**Comment by Mark-Simulacrum** [comment]
Yeah, that may have been an oversight, I'm not sure. cc @rust-lang/wg-allocators

**Comment by Amanieu** [comment]
I think this is still sound: `#[global_allocator]` must be specified on a `static`, so the type is already required to be `Sync`. And while `Global` itself is `Send`, it just just a ZST that calls through the `static`. The actual global allocator type is never moved, so `Send` doesn't apply.

The real issue here is that `#[global_allocator]` should not be allowed to be combined with `#[thread_local]`. You should use a custom allocator with the `Allocator` trait for your use case.

**Comment by repnop** [comment]
gotcha, I'll go that route then, was hoping to avoid it but hopefully shouldn't be much more work. just out of curiosity, are there technical reasons to not support it? I understand that pretty much all platforms as it stands right now wouldn't have a real way to actually make use of `!Send + !Sync` allocators since all memory is shared between threads, but I'm interested in learning about other problems it might have :)

**Comment by Amanieu** [comment]
The problem is that `Global` is `Send`, but doesn't actually hold a pointer to the `GlobalAlloc` instance since it is a ZST. This means that if you create a `Box`, send it to another thread and then free is there, the `Box` will be freed with a *different* `GlobalAlloc` instance than the one it was created with.

**Comment by repnop** [comment]
yeah okay that was my concern with the current impl, but wasn't sure if that was something that could be changed. I'll close the issue since this this shouldn't be allowed in the first place like you said. would you like me to make a new issue about preventing `#[thread_local]` and `#[global_allocator]` from being used in tandem?

**Comment by Amanieu** [comment]
Yes, please open a new issue for that.
